spec/utils/db_poller_spec.rb

```diff
@@ -0,0 +1,233 @@
+
+    after(:each) do
+      Deimos::PollInfo.delete_all
+      Widget.delete_all
+      travel_back
+    end
+
+    it 'should crash if initialized with an invalid producer' do
+      config.producer_class = 'NoProducer'
+      expect { described_class.new(config) }.to raise_error('Class NoProducer not found!')
+    end
+
+    specify '#retrieve_poll_info' do
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o5, 0o0, 0o0))
+      poller.retrieve_poll_info
+      expect(Deimos::PollInfo.count).to eq(1)
+      info = Deimos::PollInfo.last
+      expect(info.producer).to eq('MyProducer')
+      expect(info.last_sent).to eq(Time.new(0))
+    end
+
+    specify '#start' do
+      i = 0
+      expect(poller).to receive(:process_updates).twice do
+        i += 1
+        poller.stop if i == 2
+      end
+      poller.start
+    end
+
+    specify '#should_run?' do
+      Deimos::PollInfo.create!(producer: 'MyProducer',
+                               last_sent: Time.zone.local(2020, 0o5, 0o5, 0o3, 59, 0o0))
+      poller.retrieve_poll_info
+
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o5, 0o0, 0o0))
+      expect(poller.should_run?).to eq(true)
+
+      # run_every is set to 1 minute
+
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o4, 0o1, 0o0))
+      expect(poller.should_run?).to eq(true)
+
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o3, 59, 30))
+      expect(poller.should_run?).to eq(false)
+
+      travel_to(Time.zone.local(2020, 0o5, 0o5, 0o3, 58, 0o0))
+      expect(poller.should_run?).to eq(false)
+    end
+
+    describe '#process_updates' do
+      before(:each) do
+        Deimos::PollInfo.create!(producer: 'MyProducer',
+                                 last_sent: Time.zone.local(2020, 0o5, 0o4, 22, 59, 0o0))
+        travel_to(Time.zone.local(2020, 0o5, 0o5, 0o1, 0o0, 0o0))
+        allow(poller).to receive(:process_updates).and_wrap_original do |m|
+          m.call
+          poller.stop
+        end
+        allow(MyProducer).to receive(:send_events)
+        stub_const('Deimos::Utils::DbPoller::BATCH_SIZE', 3)
+
+        # old widget, earlier than window
+        Widget.create!(test_id: 'some_id', some_int: 4,
+                       updated_at: Time.local(2020, 0o5, 0o3, 23, 59, 30))
+        # new widget, before delay
+        Widget.create!(test_id: 'some_id', some_int: 10,
+                       updated_at: Time.local(2020, 0o5, 0o5, 0o0, 59, 59))
+      end
+
+      let!(:widgets) do
+        (1..7).map do |i|
+          Widget.create!(test_id: 'some_id', some_int: i,
+                         updated_at: Time.local(2020, 0o5, 0o4, 23, 59, 30 + i))
+        end
+      end
+
+      it 'should update the full table' do
+        config.full_table = true
+        expect(MyProducer).to receive(:poll_query).
+          with(Time.local(2020, 0o5, 0o4, 22, 59, 0o0),
+               Time.local(2020, 0o5, 0o5, 0o0, 59, 58),
+               :updated_at,
+               true).and_call_original
+        widgets = Widget.all
```

Comment:
Since we are comparing exact ordering, this should include a sort by id to guarantee it always matches the behaviour of find_in_batches